src/main/scala/io/druid/indexer/spark/SparkBatchIndexTask.scala

```diff
@@ -397,6 +401,28 @@ object SparkBatchIndexTask
         }
       )
 
+      log.info("Initializing emitter for metrics")
+
+      lifecycle.start()
+
+      closer.register(
+        new Closeable {
+          override def close(): Unit = lifecycle.stop()
+        }
+      )
+
+      sc.addSparkListener(new SparkListener() {
+        // Emit metrics at the end of each stage
+        override def onStageCompleted(event: SparkListenerStageCompleted): Unit = {
```

Comment:
Yeah, I think it has a chance to fire after the `closer` has fired. This `SparkListener` will process the event after the `SparkContext.stop()` gets fired. I wasn't aware that `closer.close()` closes stuff in LIFO order. I will just move `lifecycle.close` to be tied to Spark's lifecycle as you suggested. The emitter should try to flush the remaining events when it closes (in a case of `HttpPostEmitter`).